alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/actions/ChangeBiomolConcentrationInEnv.java

```diff
@@ -9,41 +9,170 @@
 import it.unibo.alchemist.model.interfaces.Reaction;
 
 /**
+ * Action implementing the changing of the concentration of a given biomolecule in environment.
  */
 public class ChangeBiomolConcentrationInEnv extends AbstractAction<Double> {
 
     private static final long serialVersionUID = -1442417685847647706L;
+    private final double delta;
+    private final Biomolecule biomolecule;
+    private final Environment<Double> env;
+    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "All provided RandomGenerator implementations are actually Serializable")
+    private final RandomGenerator rand;
 
     /**
+     * Initialize a new {@link Action} that change concentration of the given
+     * {@link Biomolecule} of a quantity {@link deltaCon}.
      * 
-     * @param node 
+     * @param node the {@link Node} where this action is located.
+     * @param biomol the {@link Biomolecule} which concentration will be changed.
+     * @param deltaCon the quantity to add to actual concentriation of {@link biomol}
+     * @param environment the {@link Environment} where the node is located.
+     * @param randomGen 
      */
-    public ChangeBiomolConcentrationInEnv(final Node<Double> node) {
+    public ChangeBiomolConcentrationInEnv(final Node<Double> node, final Biomolecule biomol, 
+            final double deltaCon, final Environment<Double> environment, final RandomGenerator randomGen) {
         super(node);
-        // TODO Auto-generated constructor stub
+        biomolecule = biomol;
+        delta = deltaCon;
+        env = environment;
+        rand = randomGen;
+    }
+
+    /**
+     * Initialize a ChangeBiomolConcentrationInEnv with delta = -1.
+     * 
+     * @param node node the {@link Node} where this action is located.
+     * @param biomol the {@link Biomolecule} which concentration will be changed.
+     * @param environment environment the {@link Environment} where the node is located.
+     * @param randomGen 
+     */
+    public ChangeBiomolConcentrationInEnv(final Node<Double> node, final Biomolecule biomol, 
+            final Environment<Double> environment, final RandomGenerator randomGen) {
+        this(node, biomol, -1, environment, randomGen);
     }
 
     @Override
     public Action<Double> cloneOnNewNode(final Node<Double> n, final Reaction<Double> r) {
-        // TODO Auto-generated method stub
-        return null;
+        return new ChangeBiomolConcentrationInEnv(n, biomolecule, env, rand);
     }
 
     @Override
     public void execute() {
-        // TODO Auto-generated method stub
-
+        // get the environment surrounding
+        final List<EnvironmentNode> environmentNodesSurrounding = getEnvironmentNodesSurrounding();
+        // if the node is an EnvironmentNode...
+        if (getNode().getClass().equals(EnvironmentNodeImpl.class)) {
+            // sort the env node randomly
+            Collections.shuffle(environmentNodesSurrounding);
```

Comment:
You don't need to shuffle. You need to pick a node randomly. Am I wrong?
What prevents you from doing something like: 
```java
final int index = random.nextInt(environmentNodesSurrounding.size());
environmentNodesSurrounding.get(index);
doWhatever();
environmentNodesSurrounding.remove(index); // if you never want to pick the same again
```